### PR TITLE
DOC-2446: New `revisionhistory_fetch_revision` option to support updating a revision when it is selected.

### PR DESCRIPTION
**Pull request by FarzadHayat**
Ticket: TINY-10362

Site: [Release Note Entry](http://docs-feature-72-doc-2446tiny-10362.staging.tiny.cloud/docs/tinymce/latest/7.2-release-notes/#revision-history)
Site: [Revision History Plugin](http://docs-feature-72-doc-2446tiny-10362.staging.tiny.cloud/docs/tinymce/latest/revisionhistory/#fetching-revisions) (changes: new Fetching Revision section, updated revisionhistory_fetch, updated revisionhistory_fetch_revision)

To test updated live demo:

1. checkout branch locally
2. change `tinymce_live_demo_url` attribute in `antora.yml` to point to `7-testing`
3. visit the live demo

Changes:
* DOC-2446: Release note entry for new `revisionhistory_fetch_revision` option
* Improve `revisionhistory_fetch` option's return data layout and change `content` field to optional
* Add `revisionhistory_fetch_revision` option and include it in `revisionhistory.adoc`
* Update Revision History live demo to include `revisionhistory_fetch_revision`
* Update Revision History demo to include more plugins and toolbar buttons

Pre-checks:
- [x] Branch prefixed with `feature/<version>/`, `hotfix/<version>/`, `staging/<version>/`, or `release/<version>/`
- [-] `modules/ROOT/nav.adoc` has been updated `(if applicable)`
- [x] Files has been included where required `(if applicable)`
- [-] Files removed have been deleted, not just excluded from the build `(if applicable)`
- [x] Files added for `New product features`, and included a `release note` entry.

Review:
- [x] Documentation Team Lead has reviewed

### CODE REVIEW

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+|===
+
+*Return data:* revision `+Object+` with the following fields:
+[cols="1,1,3",options="header"]
+|===
+|Field |Type |Required? |Description
+|`+revisionId+` | `+string+` | required | The unique string Id of the revision.
+|`+createdAt+` | `+string+` | required | A UTC datetime string in ISO-8061 format.
+|`+content+` | `+string+` | required | HTML string of the revision content.
+|===
```
Comment (tiny-ben-tran) [inline modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc:23]: Can this be a link to the revision object defined in `revisionhistory_fetch`?

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc [hunk 2]
```diff
@@ -0,0 +1,87 @@
+[[revisionhistory_fetch_revision]]
+== `revisionhistory_fetch_revision`
+
+The plugin calls this function to fetch a revision's content when a revision is selected in the revision history dialog. This function is called with the revision object that was selected in the revision history dialog. The function should return a Promise that resolves with the revision content.
```
Comment (tiny-ben-tran) [inline modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc:4]:
```suggestion
The plugin uses this option to update for when either the selected and the closest revisions have no content. The function should return a Promise that resolves with an updated revision.
```

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  revisionhistory_fetch_revision: (_editor, revision) => new Promise((resolve) => {
+	let newRevision = null;
+	for (let i = 0; i < revisions.length; i++) {
+	  const temp = revisions[i];
+	  if (temp.revisionId === revision.revisionId) {
+		  newRevision = temp;
+		  break;
+	  }
+	}
+	resolve(newRevision);
```
Comment (tiny-ben-tran) [inline modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc:84]:
```suggestion
  let newRevision = null;
  for (let i = 0; i < revisions.length; i++) {
    const temp = revisions[i];
    if (temp.revisionId === revision.revisionId) {
      newRevision = temp;
      break;
    }
  }
  resolve(newRevision);
```

**Review by tiny-ben-tran — APPROVED** [review]

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc (hunk 2 above)
Comment (kemister85) [reply modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc:4]: My thoughts, the word `should` could potentially be confusing, to me it implies it `may` or `may not` return a Promise?

File: modules/ROOT/examples/live-demos/revisionhistory/index.js [hunk 4]
```diff
@@ -82,21 +120,30 @@ const revisions = [
...
   selector: 'textarea#revisionhistory',
   height: 800,
-  plugins: 'revisionhistory',
-  toolbar: 'revisionhistory',
-  revisionhistory_fetch: get_revisions,
-  content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
-});
\ No newline at end of file
+  plugins: 'revisionhistory advlist autolink lists link image charmap preview anchor searchreplace visualblocks code fullscreen insertdatetime media table help wordcount',
+  toolbar: 'revisionhistory | undo redo | blocks | bold italic backcolor | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | removeformat | help',
```
[the remark this answers is not in the document]
Comment (FarzadHayat) [reply modules/ROOT/examples/live-demos/revisionhistory/index.js]: On second look, we don't usually include other plugins so I'll revert this.

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc (hunk 1 above)
Comment (FarzadHayat) [reply modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc]: [f6a8f86](https://github.com/tinymce/tinymce-docs/pull/3349/commits/f6a8f86a59791e76f9c2cd3be143dba38c3170c6)

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc (hunk 2 above)
Comment (FarzadHayat) [reply modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc]: [25b2863](https://github.com/tinymce/tinymce-docs/pull/3349/commits/25b28633e6cd8d91f5b5b4e4263a3622d194380c)

File: modules/ROOT/pages/7.2-release-notes.adoc [hunk 5]
```diff
@@ -70,17 +70,26 @@ In {productname} {release-version}, this issue has been resolved. Now, the slide
...
+This release of **Revision History** introduces the following addition.
 
-==== <Premium plugin name 1 change 1>
+==== New `revisionhistory_fetch_revision` option to support updating a revision.
+// #TINY-10362
 
-// CCFR here.
+In {productname} {release-version}, a new `revisionhistory_fetch_revision` option has been added to enable updating a revision after the initial fetch. If configured, when a revision is selected, the plugin uses this option to update the selected or closest revision that has no content.
+
+NOTE: As part of this addition, the `content` property is now optional. For the fetched revisions, empty string is considered as valid content.
```
Comment (kemister85) [inline modules/ROOT/pages/7.2-release-notes.adoc:84]:
```suggestion
NOTE: As part of this addition, the `content` property is now optional. For the fetched revisions, an empty string is considered as valid content.
```

File: modules/ROOT/pages/7.2-release-notes.adoc [hunk 6]
```diff
@@ -70,17 +70,26 @@ In {productname} {release-version}, this issue has been resolved. Now, the slide
...
+
+NOTE: The integrators still need to configure the `revisionhistory_fetch` option to provide the initial revisions.
```
Comment (kemister85) [inline modules/ROOT/pages/7.2-release-notes.adoc:86]:
```suggestion
IMPORTANT: The `revisionhistory_fetch` option must be configured when using the new `revisionhistory_fetch_revision` option, as it is required to provide the initial revisions.
```

File: modules/ROOT/pages/7.2-release-notes.adoc [hunk 7]
```diff
@@ -70,17 +70,26 @@ In {productname} {release-version}, this issue has been resolved. Now, the slide
...
+
+For more information on the `revisionhistory_fetch_revision` option, see: xref:revisionhistory.adoc#revisionhistory_fetch_revision[Revision History - revisionhistory_fetch_revision].
+
+This option would be helpful when dealing with revisions that have large content or preferring to only load the content when required.
```
Comment (kemister85) [inline modules/ROOT/pages/7.2-release-notes.adoc:90]:
```suggestion
This option is useful for handling revisions with large content or for loading content only when needed.
```

File: modules/ROOT/partials/configuration/revisionhistory_fetch.adoc [hunk 8]
```diff
@@ -1,18 +1,27 @@
 [[revisionhistory_fetch]]
 == `revisionhistory_fetch`
 
-The plugin uses the `revisionhistory_fetch` function to retrieve the saved versions.
+The `revisionhistory_fetch` function is used by the plugin to retrieve saved versions. This function is required for the Revision History plugin to work. The function is called when the user opens the Revision History view.
```
Comment (kemister85) [inline modules/ROOT/partials/configuration/revisionhistory_fetch.adoc:4]:
```suggestion
The `revisionhistory_fetch` function retrieves saved versions and is **required** for the Revision History plugin to work. It is called when the user opens the Revision History view.
```

File: modules/ROOT/partials/configuration/revisionhistory_fetch.adoc [hunk 9]
```diff
@@ -1,18 +1,27 @@
...
 
 The `revisionhistory_fetch` option is an asynchronous function that returns a https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise[Promise] which resolves to an array of revisions.
```
Comment (kemister85) [inline modules/ROOT/partials/configuration/revisionhistory_fetch.adoc:6]:
```suggestion
The `revisionhistory_fetch` option is an asynchronous function that returns a link:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise[Promise] which resolves to an array of revisions.
```

File: modules/ROOT/partials/configuration/revisionhistory_fetch.adoc [hunk 10]
```diff
@@ -1,18 +1,27 @@
...
 
 [NOTE]
-The {pluginname} plugin does not sort the result before displaying. This allows the integrators to sort the result according to their preference. It is recommended that the result should be sorted in reverse chronological order because it is more intuitive for the users when they can only identify a version by its date.
+The {pluginname} plugin does not sort the result before displaying. This allows integrators to sort the result according to their preference. It is recommended to sort the result in reverse chronological order, as it is more intuitive for users when they can identify a version by its date.
```
Comment (kemister85) [inline modules/ROOT/partials/configuration/revisionhistory_fetch.adoc:9]:
```suggestion
The {pluginname} plugin does not sort results before displaying them, allowing integrators to sort the results according to their preference. {companyname} recommends to sort the results in reverse chronological order, as this makes it easier for users to identify versions by date.
```

File: modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc [hunk 11]
```diff
@@ -0,0 +1,79 @@
+[[revisionhistory_fetch_revision]]
+== `revisionhistory_fetch_revision`
+When a revision is selected, the plugin uses this option to update the selected or closest revision that has no content. The function is expected to return a Promise that resolves with an updated revision.
+
+*Type:* `+Function+` (Returns a https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise[Promise])
```
Comment (kemister85) [inline modules/ROOT/partials/configuration/revisionhistory_fetch_revision.adoc:5]:
```suggestion
*Type:* `+Function+` (Returns a link:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise[Promise])
```

**Review by kemister85 — APPROVED** [review]
Copy edits and suggestions.

**Pull request merged**